lightautoml/transformers/utils.py

```diff
@@ -0,0 +1,269 @@
+
+    Note:
+        Typically is created from GroupByFactory.
+
+    Args:
+            kind (string): Id of group_by feature.
+            fit_func (function): function to calculate groups.
+            transform_func (function): function to calculate statistics based on fitted groups.
+
+    Example:
+        GroupByBase(GroupByNumDeltaMean.class_kind, GroupByNumDeltaMean.class_fit_func, GroupByNumDeltaMean.class_transform_func)
+
+    """
+
+    def __init__(self, kind, fit_func, transform_func):
+        super().__init__()
+
+        self.kind = kind
+        self.fit_func = fit_func
+        self.transform_func = transform_func
+
+        self._dict = None
+
+    def get_dict(self):
+        """Gets dict with features' statistics."""
+        return self._dict
+
+    def set_dict(self, dict):
+        """Sets dict with features' statistics."""
+        self._dict = dict
+
+    def fit(self, data, group_by_processor, feature_column):
+        """Calculate groups.
+
+        Note:
+            GroupByProcessor must be initialiaed before call to this function.
+
+        Args:
+            data (dataset): input data to extract ``feature_column``.
+            group_by_processor (GroupByProcessor): processor, containig groups.
+            feature_column (string): name of column to calculate statistics.
+
+        Returns:
+            self
+
+        """
+        assert data is not None
+        assert group_by_processor is not None
+        assert feature_column is not None
+
+        assert self.fit_func is not None
+
+        feature_values = data[feature_column].to_numpy()
+        self._dict = dict(
+            zip(
+                group_by_processor.index,
+                group_by_processor.apply(self.fit_func, feature_values),
+            )
+        )
+
+        assert self._dict is not None
+
+        return self
+
+    def transform(self, data, value):
+        """Calculate features statistics.
+
+        Note:
+            ``fit`` function must be called before ``transform``.
+
+        Args:
+            data (dataset): input data to extract ``value['group_column']`` and ``value['feature_column']``.
+            value (dict): column names.
+
+        Returns:
+            transformed data
+
+        """
+        assert data is not None
+        assert value is not None
+
+        assert self.transform_func is not None
+
+        group_values = data[value["group_column"]].to_numpy()
+        feature_values = data[value["feature_column"]].to_numpy()
+        result = self.transform_func(
+            tuple(
+                [
+                    np.nan_to_num(np.array(np.vectorize(self._dict.get)(group_values), dtype=float)),
+                    feature_values,
+                ]
+            )
+        ).reshape(-1, 1)
+
+        assert result is not None
+        return result
+
+
+class GroupByNumDeltaMean(GroupByBase):
+    """Groupby delta mean class."""
+
+    class_kind = "delta_mean"
```

Comment:
Why you are using class_kind fields to store agg action? You can store classes instead, so class_kind will not be needed.